docs/proposals/08-shoot-apiserver-via-sni.md

```diff
@@ -0,0 +1,396 @@
+    - api.<shoot>.<seed>.<domain>
+    - api.internal.<shoot>.<seed>.<domain>
+  - port:
+      number: 9443
+      name: tls
+      protocol: TLS
+    tls:
+      mode: PASSTHROUGH
+    hosts:
+    - api.internal.<shoot>.<seed>.<domain>
+---
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: kube-apiserver
+  namespace: shoot-namespace
+spec:
+  hosts:
+  - api.<shoot>.<seed>.<domain>
+  - api.internal.<shoot>.<seed>.<domain>
+  gateways:
+  - kube-apiserver-gateway
+  tls:
+  - match:
+    - port: 443
+      sniHosts:
+      - api.<shoot>.<seed>.<domain>
+      - api.internal.<shoot>.<seed>.<domain>
+    route:
+    - destination:
+        host: kube-apiserver.shoot-namespace.svc.cluster.local
+        port:
+          number: 443
+  - match:
+    - port: 9443
+      sniHosts:
+      - api.internal.<shoot>.<seed>.<domain>
+    route:
+    - destination:
+        host: proxy.shoot-namespace.svc.cluster.local
+        port:
+          number: 443
+```
+
+Chisel works by creating logical TCP connections via SSH, but it uses only one TCP connection per client/Shoot Node.
+With this approach a Chisel server (with appropriate replica count) can be deployed and it'll serve as a proxy into the cluster for those Nodes.
+
+The given that the master service ip address is `10.0.0.1`, a Pod attempting to communicate with it's API server will do the following:
+
+- `GET https://10.0.0.1/` with it's private and public keys.
+- due to the proxy's sidecar, all traffic does to the `clisel-client` listening on `10.0.0.1`.
+- `chisel-client` has already opened a TCP connection to the upstream server and it creates a new SSH channel, sending the data to `chisel-server`.
+- `chisel-server` is assigned to a specific `kube-apiserver` and sends data to it.
+
+One problem with chisel is the lack of custom certificate authority support in the client, but there is an opened PR to [fix it](https://github.com/jpillora/chisel/pull/129).
+
+> As a fun fact, the fastest proxy (in terms of overhead/latency) is based on the new [QUIC transport protocol](https://tools.ietf.org/html/draft-ietf-quic-transport-24). This protocol is used as the transport protocol for HTTP3 and supports multiplexing, multiple streams and TLSv1.3. The proxy + client is only several hundreds LoC written for this tunneling feature. The only downside is that it's UDP-based and support for UDP is lacking (different OpenStack deployments may not support it). Support for QUIC in Envoy is also [coming soon™](https://github.com/envoyproxy/envoy/issues/2557).
+
+### In-cluster communication to the apiserver when ExernalName is supported
+
+Even if in future versions of Kubernetes, the master service of type `ExternalName` is supported, we still have the problem that in-cluster workload can talk to the server via DNS. For this to work we still need the above mentioned proxy (this time listening on another IP address `10.0.0.2`). An additional change to CoreDNS would be needed:
```

Comment:
Sounds like a better bet than the side-car alternative. We have control of CoreDNS and can fully extend its config. Do you think we should push to get ExternalName in for the default or is this something that you think won't happen even if we push forward?

Reply:
It's not possible to create such a service with `spec.externalTrafficPolicy: Local`:

```
The Service "kubernetes-foo" is invalid: spec.externalTrafficPolicy: Invalid value: "Local": ExternalTrafficPolicy can only be set on NodePort and LoadBalancer service
```

Reply:
And some component have to create this VIP (dummy network interface) on every single Node